Add ERC: Cross-chain Message Format and Mailbox (#766)
ethereum/ERCs: ERCS/erc-7841.md

---
eip: 7841
title: Cross-chain Message Format and Mailbox
description: A mailbox API and message format to provide a unified cross-chain developer experience
author: Ellie Davidson (@elliedavidson), Alex Xiong (@alxiong), Philippe Camacho (@philippecamacho), and Ben Fisch (@benafisch)
discussions-to: https://ethereum-magicians.org/t/erc-7841-cross-chain-message-format-and-mailbox/22185
status: Draft
type: Standards Track
category: ERC
created: 2024-12-12

> ---

Reviewer: ```suggestion
requires: 3770
---
```

Author: This ERC doesn't require 3770.  We originally mentioned it to clarify that this ERC does not conflict with 3770.  We've removed the reference to 3770 later on the doc to hopefully clear this up.

## Abstract

> This ERC proposes a **mailbox API** and **message format** for sending and receiving data between L2s. This standard makes it easier for developers to build cross-chain applications that work over a variety of VMs, chain settlement mechanisms (e.g., ZK or optimistic settlement), and messaging protocols (e.g., synchronous or asynchronous protocols). Example applications include an intent-based bridge, a liquidity-unifying DEX operating across multiple chains, or a cross-chain lending application.

Reviewer: You abstract should sketch out a high level technical overview so readers get a sense of _how_ your proposal accomplishes its goals.

    bytes32 srcAddress;
    /// @notice The address of the recipient
    /// @dev 32 bytes are used to encode the address. In the case of an 
    ///     Ethereum address, the last 12 bytes can be padded with zeros
    bytes32 destAddress;
    /// @notice The identifier for a cross-chain interaction session
    /// @dev SHOULD be unique for every new cross-chain calls
    uint128 sessionId;
    /// @notice The message counter within an interaction session 
    /// @dev SHOULD be unique within a session
    /// @dev OPTIONAL for most asynchronous bridges where every message has a 
    ///     distinct sessionId, simply set to 0 if unused
    /// @dev E.g. In a cross-chain call: ChainA.func1 -m1-> ChainB.func2 -m2-> 
    ///     ChainC.func3 -m3-> ChainB.func4, the subscript i in m_i is the nonce
    uint128 nonce;
}

/// @title Message type
/// @notice A cross-chain message
struct Message {
    /// @notice The message metadata 
    Metadata metadata;
    /// @notice Message payload encoded using RLP serialization
    /// @dev It may be ABI-encoded function calls, info about bridged assets, 
    ///     or arbitrary message data
    bytes payload;
}
```

> It is RECOMMENDED that there exists a global rollup registry service that supports registration, deregistration, and efficient lookup of a rollup's chain ID. This work is outside the scope of this ERC, however. To display the sender and receiver of a `Message`, wallets and frontends MAY use [ERC-3770](./eip-3770.md) for chain-specific addresses.

Reviewer: > It is RECOMMENDED that there exists a global rollup registry service that supports registration, deregistration, and efficient lookup of a rollup's chain ID.

This is a weird way to word this. You're creating a recommendation not for implementers of your standard, but for the entire ecosystem.

I'd recommend something more like:

```suggestion
Implementations SHOULD use a global rollup registry service that supports registration, deregistration, and efficient lookup of a rollup's chain ID. This work is outside the scope of this ERC, however.

To display the sender and receiver of a `Message`, wallets and frontends MAY use [ERC-3770](./eip-3770.md) for chain-specific addresses.
```

Author: Updated per this suggestion!

### Mailbox APIs

Each chain SHOULD have **two canonical Mailbox contracts, one for synchronous and the other for asynchronous messaging**, responsible for managing both incoming and outgoing messages. The following APIs are RECOMMENDED to provide the minimal required functionality. Implementations of these APIs MAY include additional functions to support customized or more complex workflows.

```solidity
/// @title Mailbox contract.
/// @notice Mailbox for sending (resp. receiving) messages to (resp. from) 
///     other chains, standardized for messaging protocols that support 
///     synchronous or asynchronous, or both types of message passing.

> abstract contract Mailbox {

Reviewer: `abstract contract`s include too much implementation detail, and as such we don't generally allow them. I don't really see anything in here that couldn't be included in an interface, unless the storage requirements of `inbox` are meaningful?

Reviewer 2: > unless the storage requirements of inbox are meaningful?

Yes, we find this detail important and a deliberate design decision, as we explained in the very next paragraph: 
> The concrete data structure used to store the `inbox` queue SHOULD be a hash-map-like `mapping` in Solidity to enable efficient lookup by the dApps with payload-independent query keys. Being able to read/receive messages based on their metadata only, rather than their actual payload, is critical in achieving synchronous messaging with a dynamic payload known only at runtime.

that said, we are happy to comment this out and leave as comment/hints to the implementors, and convert this to an `interface` instead.

hopefully with our explanation and reference implementation, it should be clear.

Reviewer: You can absolutely leave hints to the implementers, but the requirements should only describe externally visible behavior. If some insane developer wants to binary search an array or use a B-Tree, that's up to them.

As mentioned above, pre-filling the inbox of the destination chain incurs additional gas costs which are manageable on L2s. We list some advantages of our approach:

- Eliminates Merkle inclusion proof verification when receiving every message. Some messaging protocols obviate Merkle proofs completely, even during `populateInbox()`, and delay the mailbox check to the settlement layer.
- Allows upgrade of the coordination protocol (e.g. the internal of `populateInbox()`and/or the settlement layer logic) without requiring changes to connected apps.
- Enables user-signed transactions on the destination chain instead of shared sequencer-signed transactions in synchronous messaging. This flexibility simplifies gas payment handling and ensures a consistent `msg.sender`. Detailed explanations are omitted for brevity (extended answers on the website).

### Related Proposals

> In comparison to IBC-like standards, this ERC is designed to work in a *stateless* manner. Messages do not need to pass a proof from the source chain at the time they are consumed on the destination chain. This allows use cases such as synchronous composability and intra-block messaging since messages don’t need to include finalized state from the source chain. Additionally, this ERC does not require multiple steps to establish a link between two chains. Messages can be directly sent from one chain to another in a single step.

Reviewer: > IBC-like

Please expand uncommon abbreviations the first time they are used.

Author: Fixed!

> [ERC-7683](./eip-7683.md) standardizes intent-based systems by defining structs for orders and interfaces for settlement smart contracts. This standard is application-specific and aimed at designers of cross-chain intent systems, while our proposal is more general and targets developers implementing arbitrary cross-chain applications. However, an intent system based on ERC-7683 **can be built on top** of our standard due to its modularity. An application implementing ERC-7683 could use the `Mailbox` API defined in this proposal to send `originData` from event messages between the source chain (where user funds are deposited) and the destination chain(s) (where intents are solved). We provide more details in the *Example Usage* section.

Reviewer: If you mention ERC-7683, your standard will not be able to progress past its status. See [`markdown-link-status`](https://ethereum.github.io/eipw/markdown-link-status/) for more detail.

There's absolutely nothing wrong with that. I just wanted to let you know.

Reviewer 2: So should we just mention it literally but not link it (or establish any dependency relationship)?

Reviewer: Mentioning it at all will create the implicit dependency.

### Example Usage - Sync and Async Cross-chain Transfers

> We show a *possible* implementation of the mailbox contract for synchronous messaging protocol, and explain how it can be easily modified to support asynchronous protocols as well.

Reviewer: This sounds like it belongs in the reference implementation section?

Reviewer 2: yes we will rename them as "reference implementation", and separate out the transfer section as its individual "example usage" subsection.

    }
}

/// Contract deployed on chain B
/// We assume this contract is deployed at the address FOO_CONTRACT_ADDRESS
/// This contract has a function that is called from chain A
contract Foo {
    function fun(uint256 parameter) public {
        require(parameter == 42);
    }
}
```

## Backwards Compatibility

No backward compatibility issues found. Since this is an opt-in protocol, L2s that do not opt-in to this will not be affected. Furthermore, this protocol can operate in existing cross-chain flows today, such as in intent-based bridges or native bridging of assets between L2s.

## Security Considerations

Security concerns for the messaging format and mailbox APIs themselves are minimal, as the protocol specification focuses on providing a rich and expressive interface for various message passing protocols. Security responsibilities lie with the underlying messaging protocol design and the application utilizing it. This specification ensures the interfaces are flexible enough to support the majority of cross-chain messaging protocols, while security within those protocols is outside the scope of this proposal.

## Copyright

> Copyright and related rights waived via CC0.

Author: Fixed!

    uint32 destChainId;
    /// @notice The address of the sending party
    /// @dev 32 bytes are used to encode the address. In the case of an
    ///     Ethereum address, the last 12 bytes can be padded with zeros
    bytes32 srcAddress;
    /// @notice The address of the recipient
    /// @dev 32 bytes are used to encode the address. In the case of an 
    ///     Ethereum address, the last 12 bytes can be padded with zeros
    bytes32 destAddress;
    /// @notice The identifier for a cross-chain interaction session
    /// @dev SHOULD be unique for every new cross-chain calls
    uint128 sessionId;
    /// @notice The message counter within an interaction session 
    /// @dev SHOULD be unique within a session
    /// @dev OPTIONAL for most asynchronous bridges where every message has a 
    ///     distinct sessionId, simply set to 0 if unused
    /// @dev E.g. In a cross-chain call: ChainA.func1 -m1-> ChainB.func2 -m2-> 
    ///     ChainC.func3 -m3-> ChainB.func4, the subscript i in m_i is the nonce
    uint128 nonce;
}

/// @title Message type
/// @notice A cross-chain message
struct Message {
    /// @notice The message metadata 
    Metadata metadata;

>     /// @notice Message payload encoded using RLP serialization

Reviewer 3: It seems that `payload` can use any encoding, not just RLP. (The two examples both use ABI encoding.)

Reviewer 2: Yes, thanks for catching that. we should remove the mentioning of RLP here.
Theoreticall yes you can use any encoding, but in practice on EVM chain, most likely just ABI encoding.

### Use Metadata Digest Instead of `sessionId` for Message Query Key

For message lookups in the mailbox, the query key is derived from the hash of all message metadata rather than relying on a single `sessionId` field. This is because the `sessionId` derivation is customizable and may not adequately bind to key metadata like source and destination addresses. In contrast, the wrapping messaging protocol may enforce permissions for sending or receiving based on these metadata fields, so the query key must bind to the entire set of metadata.

Observe that when applications allow users to define `sessionId`, these values may not be unique across messages. Depending on the implementation of the inbox, this could be a concern. For example,  a mapping-based inbox needs an additional *nullifier set* to enforce the uniqueness of message metadata and avoid message overwrites in the case of a colliding map-key.

> ### Gas Cost

Reviewer: This section still refers to a particular implementation. It should go in your reference implementation section.

                    uid
                )
            );
    }

    /// @notice Conceptual "cleanup/reset" of mailbox after each block since 
    ///     sync msgs are received immediately.
    function _resetMailbox() private {
        delete inbox[block.number - 1];
        delete inboxDigest[block.number - 1];
        delete outboxDigest[block.number - 1];
    }

    /// @notice Send a message to another chain
    function send(
        uint32 destChainId,
        address destAddress,
        uint256 uid,
        bytes memory payload
    ) public {
        bytes32 key = getMetadataDigest(
            this.chain_id(),
            destChainId,
            bytes32(srcAddress),
            bytes32(msg.sender),
            uid
        );

        // Prevent overwriting the same key
        require(!outboxNullifier[key]);
        outboxNullifier[key] = true;

        // Update the outbox digest
        // digest' = H(digest | metadata | payload)
        outboxDigest[block.number][this.chain_id()] = keccak256(
            abi.encodePacked(
                outboxDigest[block.number][this.chain_id()],
                key,
                m.payload
            )
        );
    }

>     /// @dev This function is called by the Coordinator. It can only be called 

Reviewer 3: `Coordinator` is not defined in this doc.

Reviewer 2: We did define the "coordination protocol" (basically the off-chain mechanism that relay the messages) in our terminology section. While it's true we didn't explicitly define its role, it's a bit out-of-scope to fully and accurately describe what a coordinator does. 

that said, I agree maybe it doesn't hurt to add a simple description like: 
"coordinators are the entity that monitors the sending events on multiple rollups, determines their message payload and relay them to the destination by populating them into the correct inbox"

Reviewer 2: on a side note, @philippecamacho do you think we should remove this comment/restriction. Didn't we discuss internally that we don't have to restrict the permission to call this function, as long as it's called once (by anyone). It should be fine?

Reviewer 4: Yes, because the _coordinator_  builds the block so it is enough to ensure this function can only be called once per block.
